Add proxy waiters to allow the Stream to trigger AsyncWrite oper…

sdroege opened the pull request:
…ations

As a side effect also gets rid of unsafe code and raw pointers.

----

* [x] Needs some cleanup and better documentation
* [x] Needs testing if it actually solves the problem
* [x] Is it a problem that we never clear the wakers from the `AtomicWaker`s? It might cause spurious wakeups to a task that is not interested in us anymore, so maybe makes sense to clear whenever a non-`Pending` result is returned.

dbcfd committed:
Updates for tokio master

dbcfd committed:
More logging

dbcfd committed:
Remove trace on start send

sdroege committed:
Add proxy waiters to allow the Stream to trigger AsyncWrite operations

As a side effect also gets rid of unsafe code and raw pointers.

We have the problem that external read operations (i.e. the Stream impl)
can trigger both read (AsyncRead) and write (AsyncWrite) operations on
the underyling stream. At the same time write operations (i.e. the Sink
impl) can trigger write operations (AsyncWrite) too.
Both the Stream and the Sink can be used on two different tasks, but it
is required that AsyncRead and AsyncWrite are only ever used by a single
task (or better: with a single waker) at a time.

Doing otherwise would cause only the latest waker to be remembered, so
in our case either the Stream or the Sink impl would potentially wait
forever to be woken up because only the other one would've been woken
up.

To solve this we implement a waker proxy that has two slots (one for
read, one for write) to store wakers. One waker proxy is always passed
to the AsyncRead, the other to AsyncWrite so that they will only ever
have to store a single waker, but internally we dispatch any wakeups to
up to two actual wakers (one from the Sink impl and one from the Stream
impl).

sdroege committed:
Update to tokio 0.2 release

sdroege commented:
Confirmed that it solves the problem from the testcase at least. And updated to tokio 0.2 release while at it.

Now to clean things up and fix the too long storage of the wakers.

sdroege committed:
Run travis with stable Rust toolchain again

sdroege committed:
Remove WebSocketStream::send()

The same functionality is already provided via StreamExt::send() and
unlike the custom implementation it handles WouldBlock correctly and not
as an error.

sdroege committed:
Remove custom WebSocketStream::close() implementation

Instead simply send an owned Closed message. This simplifies the code
and among other things also handles errors like WouldBlock correctly
instead of handling them like a real error.

sdroege commented:
> Is it a problem that we never clear the wakers from the AtomicWakers? It might cause spurious wakeups to a task that is not interested in us anymore, so maybe makes sense to clear whenever a non-Pending result is returned.

This is actually fine. The `AtomicWaker` is only waking exactly once and then the stored waker is reset. So at most we would once unnecessarily wake a task, but even that seems very unlikely.

I'd keep it as-is instead of complicating the code more (each caller would have to check the return value and reset it if not `Pending`).

sdroege commented:
This should be good to go now. The original testcase for the problem runs fine here now, and I've added the last two commits because that testcase would otherwise sometimes error out with `WouldBlock` (instead of it causing a `Pending` and suspending the future).

Please review and let me know what you think!

sdroege committed:
Update dependencies

sdroege committed:
Remove unused #[pin_project] attribute

sdroege committed:
Use #[pin_project] in stream implementation to get rid of remaining unsafe code

Now the whole crate has no unsafe code left.

dunnock commented:
hi @sdroege , for Cargo.toml what do you think of cherry picking only required features for dependencies, like in this diff:
https://github.com/dbcfd/tokio-tungstenite/pull/7/files#diff-80398c5faae3c069e4e6aa2ed11b28c0

sdroege committed:
Depend only on the necessary tokio features instead of "full"

Reduces the number of dependencies and code to compile.

Based on a PR by Maksym Vorobiov.

sdroege commented:
> for Cargo.toml what do you think of cherry picking only required features for dependencies, like in this diff:

@dunnock Thanks, I've integrated a commit like that.

dbcfd commented on src/compat.rs:
```diff
@@ -3,55 +3,125 @@ use std::io::{Read, Write};
     S: Unpin,
 {
-    fn with_context<F, R>(&mut self, f: F) -> Poll<std::io::Result<R>>
+    fn with_context<F, R>(&mut self, read: bool, f: F) -> Poll<std::io::Result<R>>
```
What if we use an enum here so it is a little more explicit?

```rust
enum ContextWaker {
  Read,
  Write,
}

impl ContextWaker {
  pub fn waker<S>(&self, allow_std: &AllowStd<S>) -> &task::AtomicWaker {
    match self {
      Read => &allow_std.read_waker_proxy,
      Write => &allow_std.write_walker_proxy,
    }
  }
}
```

sdroege replied:
Done (slightly different, the above also won't work because then the return value keeps `allow_std` borrowed and it can't be mutably borrowed anymore for the `Pin`)

dbcfd commented on src/compat.rs:
```diff
@@ -121,7 +191,10 @@ where
         Ok(v) => Poll::Ready(Ok(v)),
-        Err(WsError::Io(ref e)) if e.kind() == std::io::ErrorKind::WouldBlock => Poll::Pending,
+        Err(WsError::Io(ref e)) if e.kind() == std::io::ErrorKind::WouldBlock => {
+            debug!("WouldBlock");
```
Trace level?

sdroege replied:
Indeed

dbcfd commented on src/lib.rs:
```diff
@@ -214,19 +210,25 @@ impl<S> WebSocketStream<S> {
-        };
-        let mut g = compat::Guard(&mut self.inner);
-        f(&mut (g.0))
+        if let Some((read, ctx)) = ctx {
```
Can expand on the read/write enum here.

sdroege replied:
Done

dbcfd reviewed (commented):
In addition to the comments, I'm currently holding off on merging this till hyper gets another release, since hyper is pinned to 0.2.0-alpha.6

sdroege commented:
> In addition to the comments, I'm currently holding off on merging this till hyper gets another release, since hyper is pinned to 0.2.0-alpha.6

Thanks for the review. I've updated it accordingly.

sdroege committed:
Remove unused bytes dependency

dbcfd commented:
@sdroege New plan :D Going to merge this in, create a 0.10.0 tag. Then I will downgrade dependencies on master (so anyone using this will still remain working) and create a 0.10.0-alpha.1 tag.

dbcfd commented:
0.10.0-alpha.2 targetting tokio 0.2.0-alpha.6: https://github.com/dbcfd/tokio-tungstenite/releases/tag/v0.10.0-alpha.2
0.10.0 targeting tokio 0.2: https://github.com/dbcfd/tokio-tungstenite/releases/tag/v0.10.0

sdroege commented:
Thanks for merging :) A bit sad though that you merged it all in a single commit, I tried to keep each individual change in a separate commit for having an easier to read commit history and also to keep the names of the authors of some of the commits not by me.
